OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,285 @@
+            value_array.upd().push_back(SimTK::Value<ValueType>{value});
+            _dictionary.setValueArrayForKey(key, value_array);
+        }
+
+        void removeValueForKey(const std::string& key) {
+            _dictionary.removeValueArrayForKey(key);
+        }
+
+        std::vector<std::string> getKeys() const {
+            return _dictionary.getKeys();
+        }
+
+    private:
+        ValueArrayDictionary _dictionary;
+    }; // ValueDictionary
+
+public:
+    using TableMetaData       = ValueDictionary;
+    using DependentsMetaData  = ValueArrayDictionary;
+    using IndependentMetaData = ValueDictionary;
+
+    AbstractDataTable()                                      = default;
+    AbstractDataTable(const AbstractDataTable&)              = default;
+    AbstractDataTable(AbstractDataTable&&)                   = default;
+    AbstractDataTable& operator=(const AbstractDataTable&)   = default;
+    AbstractDataTable& operator=(AbstractDataTable&&)        = default;
+    virtual std::unique_ptr<AbstractDataTable> clone() const = 0;
+    virtual ~AbstractDataTable()                             = default;
+
+    /** Get metadata associated with the table.                               */
+    const TableMetaData& getTableMetaData() const {
+        return _tableMetaData;
+    }
+
+    /** Update metadata associated with the table.                            */
+    TableMetaData& updTableMetaData() {
+        return _tableMetaData;
+    }
+
+    /** Get number of rows.                                                   */
+    size_t getNumRows() const {
+        return computeNumRows();
+    }
+
+    /** Get number of columns.                                                */
+    size_t getNumColumns() const {
+        return computeNumColumns();
+    }
+    
+    /** Get metadata associated with the independent column.                  */
+    const IndependentMetaData& getIndependentMetaData() const {
+        return _independentMetaData;
+    }
+    
+    /** Set metadata associated with the independent column.                  */
+    void 
+    setIndependentMetaData(const IndependentMetaData& independentMetaData) {
+        _independentMetaData = independentMetaData;
+        validateIndependentMetaData();
+    }
+
+    /** Get metadata associated with the dependent columns.                   */
+    const DependentsMetaData& getDependentsMetaData() const {
+        return _dependentsMetaData;
+    }
+    
+    /** Set metadata associated with the dependent columns.                   */
+    void 
+    setDependentsMetaData(const DependentsMetaData& dependentsMetaData) {
+        _dependentsMetaData = dependentsMetaData;
+        validateDependentsMetaData();
+    }
+
+protected:
+    virtual size_t computeNumRows() const            = 0;
+    virtual size_t computeNumColumns() const         = 0;
+    virtual void validateIndependentMetaData() const = 0;
+    virtual void validateDependentsMetaData() const  = 0;
```

Comment:
> Should any of these be extendValidate, etc?

I think we should reserve the `extend` prefix for virtuals where the implementor is required to invoke its base class method before adding their own functionality. For pure virtuals or virtuals where the override is expected to replace the base implementation in its entirety, "extend" isn't really the right word.